# 5.6 try语句块和异常处理

**异常**指存在于运行时的反常行为，这些行为超出了函数正常功能的范围。

**异常处理机制**为程序中**异常检测**和**异常处理**这两部分的协作提供支持。

异常处理包括：

- throw表达式（throw expression）
- try语句块（try block）
- 异常类（exception class）



### 5.6.1 throw表达式

程序的异常检测部分使用throw表达式**引发**（raise）一个异常。

throw表达式包含**关键字throw和紧随其后的一个表达式**，其中表达式的类型就是抛出的异常类型。后面通常紧跟一个分号，从而构成一条表达式语句。

抛出异常将终止当前的函数，并把控制权转移给能处理该异常的代码。

```cpp
//例：如果ISBN不一样就抛出一个异常，该异常是类型runtime_error的对象。
if(item1.isbn() != item2.isbn())
    throw runtime_error("Data must refer to same ISBN");
cout << item1 + item2 << endl;
```

例中类型runtime_error是标准库异常类型的一种，定义在stdexcept头文件中。我们必须初始化runtime_error的对象，方式是给它提供一个string对象或者一个C风格的字符串，这个字符串中有一些关于异常的辅助信息。



### 5.6.2 try语句块

**语法形式**：

```cpp
try{
    program-statements
}catch(exception-declaration){
    handler-statements
}catch(exception-declaration){
    handler-statements
}//...
```

try语句块以关键字try开头，随后紧跟一个块。try块之后是一个或多个catch子句。

catch子句包括三部分：关键字catch、**异常声明**（exception declaration）以及一个块。

program-statements组成程序的正常逻辑，其中可以有包括声明在内的任意C++语句。

**执行过程：**当选中了某个catch子句处理异常之后，执行与之对应的块，catch一旦完成，程序跳转到try语句块最后一个catch子句之后的那条语句继续执行。

```cpp
//例：
while(cin >> item1 >> item2){
    try{
        //执行添加两个Sales_item对象的代码
        //如果添加失败，代码抛出一个runtime_error异常
    }catch(runtime_error err){
        //提醒用户两个ISBN必须一致，询问是否重新输入
        cout << err.what()
             << "\nTry Again? Enter y or n" << endl;
        char c;
        cin >> c;
        if(!cin || c == 'n')
            break;//跳出while循环
    }
}
```

其中，runtime_error的what成员返回的是初始化一个具体对象时所用的string对象的副本，如果上一节编写的代码抛出异常，则本节的catch子句输出：

```cpp
Data must refer to same ISBN
Try Again? Enter y or n
```

**注意**：

- try语句块内声明的变量在块外部无法访问，特别是在catch子句内也无法访问。



#### 函数在寻找处理代码的过程中退出

在复杂系统中，程序在遇到抛出异常的代码前，其执行路径可能已经经过了多个try语句块。

寻找处理代码的过程与函数调用链刚好相反。当异常被抛出时，首先搜索抛出该异常的函数。如果没有找到匹配的catch子句，终止该函数，并在调用该函数的函数中继续寻找。如果还是没有找到匹配的catch子句，这个新函数也被终止，继续搜索调用它的函数。以此类推，沿着程序的执行路径逐层回退，直到找到适当类型的catch子句为止。

如果最终还是没能找到任何匹配的catch子句，程序转到名为terminate的标准库函数。该函数的行为与系统有关，一般情况下，执行该函数将导致程序非正常退出。

对于没有任何try语句块定义的异常，也按照类似的方式处理，即调用terminate函数并终止当前程序的执行。



### 5.6.3 标准异常

C++标准库定义了一组类，用于**报告标准库函数遇到的问题**。这些异常类可以在用户编写的程序中使用，他们分别定义在四个头文件中：

- **exception头文件**：定义了最通用的异常类exception。它只报告异常的发生，不提供任何额外信息。
- **stdexcept头文件**：定义的异常类见表5.1。其中包括runtime_error（只有在运行时才能检测出的问题）。
- **new头文件**：定义了bad_alloc异常类，将在12.1.2节介绍
- **type_info头文件**：定义了bad_cast异常类，将在19.2节介绍。

**初始化**：

- exception、bad_alloc、bad_cast对象只能以默认初始化的方式初始化，不允许提供初始值。其他异常类型的行为恰好相反，不允许默认初始化，必须提供初始值（应使用string对象或C风格字符串初始化），该初始值含有错误相关的信息。

**成员函数**：

异常类型只定义了一个名为**what**的成员函数。what函数**没有参数**，返回值是C风格字符串（即const char*）。该字符串的目的是**提供关于异常的一些文本信息**。

what返回的内容：如果异常类型有一个字符串初始值，则返回该字符串；反之，返回内容由编译器决定。

